File: projects/03-patient-vitals/fog/aggregation.test.js
"use strict";

const test = require("node:test");
const assert = require("node:assert/strict");
const { summarizeWindow } = require("./aggregation");

const READINGS = [
  { ts: "t0", value: 60.0 },
  { ts: "t1", value: 70.0 },
  { ts: "t2", value: 80.0 },
];

test("summarizeWindow computes basic stats", () => {
  const s = summarizeWindow("heart_rate", "patient-1", "bpm", READINGS, "start", "end");
  assert.equal(s.count, 3);
  assert.equal(s.min, 60.0);
  assert.equal(s.max, 80.0);
  assert.equal(s.avg, 70.0);
  assert.equal(s.latest, 80.0);
});

test("summarizeWindow carries metadata", () => {
  const s = summarizeWindow("spo2", "patient-9", "%", READINGS, "s", "e");
  assert.equal(s.sensor_type, "spo2");
  assert.equal(s.site_id, "patient-9");
  assert.equal(s.unit, "%");
  assert.equal(s.window_start, "s");
  assert.equal(s.window_end, "e");
});

test("latest is the last reading", () => {
  const readings = [{ ts: "t0", value: 5.0 }, { ts: "t1", value: 7.5 }];
  assert.equal(summarizeWindow("systolic_bp", "p", "mmHg", readings, "s", "e").latest, 7.5);
});

File: projects/02-industrial-equipment/fog/src/test/java/com/fec/industrial/fog/AggregationTest.java
package com.fec.industrial.fog;

import org.junit.jupiter.api.Test;

import java.util.List;

import static org.junit.jupiter.api.Assertions.*;

class AggregationTest {

    static final List<Reading> READINGS = List.of(
        new Reading("t0", 10.0),
        new Reading("t1", 20.0),
        new Reading("t2", 30.0)
    );

    @Test
    void rollUpBasicStats() {
        Aggregation.Summary s = Aggregation.condenseWindow("motor_temperature", "line-1", "C", READINGS, "start", "end");
        assertEquals(3, s.count());
        assertEquals(10.0, s.min());
        assertEquals(30.0, s.max());
        assertEquals(20.0, s.avg());
        assertEquals(30.0, s.latest());
    }

    @Test
    void rollUpCarriesMetadata() {
        Aggregation.Summary s = Aggregation.condenseWindow("vibration", "line-7", "mm/s", READINGS, "s", "e");
        assertEquals("vibration", s.sensorType());
        assertEquals("line-7", s.siteId());
        assertEquals("mm/s", s.unit());
        assertEquals("s", s.windowStart());
        assertEquals("e", s.windowEnd());
    }

    @Test
    void latestIsLastReading() {
        List<Reading> readings = List.of(new Reading("t0", 5.0), new Reading("t1", 7.5));
        assertEquals(7.5, Aggregation.condenseWindow("power_draw", "l", "kW", readings, "s", "e").latest());
    }

    @Test
    void singleReadingWindowCollapsesToThatValue() {
        Aggregation.Summary s = Aggregation.condenseWindow("bearing_acoustic", "line-1", "dB",
            List.of(new Reading("t0", 72.0)), "s", "e");
        assertEquals(1, s.count());
        assertEquals(72.0, s.min());
        assertEquals(72.0, s.max());
        assertEquals(72.0, s.avg());
        assertEquals(72.0, s.latest());
    }

    @Test
    void averageIsRoundedToThreeDecimalPlaces() {
        // 1 + 2 + 2 = 5, /3 = 1.6666..., rounded to 1.667
        List<Reading> readings = List.of(new Reading("t0", 1.0), new Reading("t1", 2.0), new Reading("t2", 2.0));
        assertEquals(1.667, Aggregation.condenseWindow("vibration", "l", "mm/s", readings, "s", "e").avg());
    }

    @Test
    void minAndMaxTrackExtremesRegardlessOfArrivalOrder() {
        List<Reading> readings = List.of(new Reading("t0", 30.0), new Reading("t1", 10.0), new Reading("t2", 20.0));
        Aggregation.Summary s = Aggregation.condenseWindow("motor_temperature", "l", "C", readings, "s", "e");
        assertEquals(10.0, s.min());
        assertEquals(30.0, s.max());
        assertEquals(20.0, s.latest(), "latest follows arrival order, not magnitude");
    }
}
